setup.py

```diff
@@ -974,8 +974,53 @@ def print_box(msg):
         print('|{}{}|'.format(l, ' ' * (size - len(l))))
     print('-' * (size + 2))
 
+def pre_build_checks():
+    if sys.version_info >= (3, 11):
+        # Dynamo is using internal data structures from CPython.
+        # We ensure here that there were no ABI changes in the CPython
+        # version being used to compile PyTorch.
+        def get_py_interpreter_struct(file):
+            res = []
+            record = False
+            with open(file) as f:
+                for l in f:
+                    if "typedef struct _PyInterpreterFrame" in l:
+                        record = True
+                    if record:
+                        res.append(l)
+                    if "} _PyInterpreterFrame;" in l:
+                        break
+            if not res:
+                raise RuntimeError(f"Could not find the _PyInterpreterFrame struct in {file}")
+            return res
+
+        python_header = os.path.join(cmake_python_include_dir, "internal", "pycore_frame.h")
+        pytorch_types = os.path.join(cwd, "torch", "_dynamo", "types.py")
+
+        if not os.path.exists(python_header):
+            raise RuntimeError("Could not find CPython internal headers which are required "
+                               "for building PyTorch 3.11+. Make sure to have CPython properly "
+                               "installed and if this header was removed in a recent version of "
+                               "CPython, update PyTorch to support it.")
+
+        if not os.path.exists(pytorch_types):
+            raise RuntimeError("Could not find the dynamo.types module. If you removed it, please "
+                               "update the build script to point to the new location of the _PyInterpreterFrame struct")
+
+        python_struct = get_py_interpreter_struct(python_header)
+        pytorch_struct = get_py_interpreter_struct(pytorch_types)
```

Comment:
Hmm instead of parsing the headers, wouldn't it be more convenient to just compile same test code against two headers and make sure they yield same results?